### ISSUE TITLE
AMD/OOP/Require-JS port PR

### ISSUE THREAD

**Issue by gbaumgart** [WebCabin/wcDocker#65]
Hi,
I know you're little busy(which is good) but I wanted to keep track of the mentioned PR here. See below the changes/steps I will apply  to a fresh fork(now):

1. adding [dcl](https://github.com/uhop/dcl) & require-js(AMD) as bower dependencies

2. converting all classes to dcl classes. Here is an example library from IBM: https://github.com/ibm-js/deliteful

This step is pretty big, and also impacts the build process.

3. Export dcl classes back to global so that wcDocker classes are in global again.

4. Create grunt tasks for final build

6. Create grunt-task for generating docs (I will use here some IBM stuff too since its similar to my and your existing stuff too)

7. Ready for review. 

I will will do step 1. til 4. today. It would be great you're around. I am not familar with PRs as such and I'd like to avoid ongoing merges of your future updates. 

Ok, I will start now and send you a few updates here today as well.

**Comment by mc007** [comment]
1. updated bower.json : https://github.com/mc007/wcDocker/commit/259a53f1a5aba93d5c64c9e9bee249a990fce3cc

1.1 added copy of indexDev.html & demo.js to as starting point to begin step 2. : https://github.com/mc007/wcDocker/commit/dde1af69c3eaf398308fadf3ffc119c7270e6231

**Comment by mc007** [comment]
1.2 all files in Code/ cleaned from unneeded commas and semicolons, intendeds reformatted.

**Comment by mc007** [comment]
***1.3 until here***:
 https://github.com/mc007/wcDocker/commit/2d3a6f283365d316efabe9c3ae64ff9068fab5d8 nothing really changed, just code cleanup and dependencies updated.

**Comment by mc007** [comment]
1.4 
- Basic conversion to dcl modules/classes complete, until here: https://github.com/mc007/wcDocker/commit/82eaafd1e02cffad917c785a9dcf0bc157873ed6

- moved redundant function class#docker() into a common shared base class : Code/base.js

**Comment by mc007** [comment]
1.5 
- AMD - Port complete and fully working, see indexDevAMD.html (requires Code/samples/demoAMD.js)
- theme builder not yet integrated
- "instanceof wcDocker" calls replaced with internal instanceOf function (in Code/base)

https://github.com/mc007/wcDocker/commit/96fe528e6670c603611d67812aad231f0dd7f2c7

**Comment by mc007** [comment]
1.6 Ported theme-builder, all working as before

https://github.com/mc007/wcDocker/commit/634d1a62d3f56bceb2537cffb31acda61f6c2ce0

Next: grunt-tasks && adjusting jsDoc for dcl classes

**Comment by Lochemage** [comment]
@mc007 

Looks like you've done quite a bit of work getting these changes ready to go! To make a pull request, go to your fork on Github, enter the pull request section from the menu on the right, and then click the green "New pull request" button.

**Comment by mc007** [comment]
hi, thanks. i just can't get the wcIFrame working (crashes at __onMoved because this.docker() seems not to work here, in the other classes it does). 

Ok, still need the grunt-tasks for docs & build and then I do the PR.

**Comment by Lochemage** [comment]
@mc007 

Ah, okay.  Getting the iFrame working is of course important, I'll be happy to help with that if I can.

**Comment by mc007** [comment]
sure, 
1. just clone my fork (https://github.com/mc007/wcDocker)
2. then bower install
3. then open wcDocker/indexDevAMD.html

you will see in the console: "have no docker". its defined in Code/base and works for all others. 

sorry, have really no idea why. 

Another problem I came around whilst creating a grunt task for the doc build is: 
1. since its dcl based, jsdoc has problems to parse it right (still trying)
2. I actually know that problem, and I used the same approach and node modules like IBM. 
3. the solution: instead of having all 'classes' in 'Classes' within the API docs, its all under "Modules" : 

![image](https://cloud.githubusercontent.com/assets/1205100/9612305/22d6c70e-50e5-11e5-8f5e-7d57c0f980b4.png)


I am not sure its a problem or its as intended. However, I hope you don't have a problem with the difference, else I am really screwed ;-)

**Comment by Lochemage** [comment]
@mc007 

Alright, I'm starting to go through the changes you have made so far. Is there any way to make it so the original `index.html` and `indexDev.html` work?  The changes you have made for `indexDevAMD.html` pretty much forces all users to change the way they initialize their own code (and also forces them to include dcl).

I don't mind having to use dcl during development when updating the project, but for those that don't want to use that library in their own project, there should be some other solution; perhaps the compiler could also generate non-dcl build files that somehow convert the dcl modules back into global variables?

**Comment by mc007** [comment]
index.html = sure, indexDevHTML was just for me as transition. I will remove also other files when I am done.
dcl/require.js = at the end the build should produce a minified file as before (includes dcl/require.js). 

So, I am not sure what you mean now, dcl/require-js will be always there, its just hidden to users who don't use the dev version, so they don't have any API changes (except they can use Dcl's OOP now). 
However, I can check for a non/optional-dcl version but you loose the strength of it, eg: you can't remove redundant code anymore, etc....

**Comment by Lochemage** [comment]
@mc007 

All I mean is that a user should not have to initialize docker like this:

    <script src="bower_components/requirejs/require.js"></script>
    <script type="text/javascript">
        require.config({
            baseUrl: "./bower_components",
            packages: [
                {
                    name: 'dcl',
                    location: './dcl'   //points to bower_components/dcl
                },
                {
                    name: 'wcDocker',
                    location: '../Code'
                }
            ]
        });
    </script>
    <script type="text/javascript">
       require.config({
            config: {
                'wcDocker/samples/demoAMD':{
                    //  enable/disable theme builder
                    themeBuilder: true,
                    //  enable/disable tutorials
                    tutorials:false
                }
            }
        });
        try {
            //require demo
            require([
                "wcDocker/samples/demoAMD"
            ], function () {
            });
        }catch(e){
            console.error('error loading demo ' + e , e);
            console.dir(e.stack);
        }
    </script>
    define(["../docker", "../tabframe", "../iframe", "wcDocker/ThemeBuilder", "module"],
      function (wcDocker, wcTabFrame, wcIFrame, wcThemeBuilder, module) {
        var myDocker = new wcDocker('.dockerContainer', {
          allowDrawers: true,
          responseRate: 10
        });
      }
    );

This requires everyone to add quite a lot to their code just to get wcDocker working again. I don't mind maintaining a version of wcDocker with dcl, but there needs to be a built version that still behaves like it used to, with the classes being global and accessible without loading them like a module.

**Comment by mc007** [comment]
sure, i am on it :-)

**Comment by mc007** [comment]
most is done: https://github.com/mc007/wcDocker/blob/master/ChangeLogAMD.md

Still the iframe issue to solve....if you could help, that would be great.

**Comment by mc007** [comment]
Is there still anything you want me to change? I am heading now over to the open issues, jsdoc as grunt and the iframe bug. I also can still do something about dcl. Just ask and its given :-)
g

Btw, I just found your wcPlay lib; that is something I am looking since long for, "NowFlow" or "Node-Red" was way too heavy. I also tried to do my own graph-lib but gave up very fast. 
However, looks you got the thing right: inputs/terminals for data and triggers! awesome. I will check it out this days more in detail. If you like I can apply same changes to that lib, that would enable me to work on it :-)

g

**Comment by Lochemage** [comment]
@mc007 

Sorry, I've pulled the latest code from your repo and saw there were a lot of new changes, but I haven't had the time to really look at it yet. On top of that, I'm getting over a bad flu, so that doesn't make it any easier.

I'll try to get to it soon, though!

**Comment by mc007** [comment]
yeah, same here, since 2 days stuck with lemon&honey tee

**Comment by mc007** [comment]
btw. nothing really 'changed' per se, its just different packaged and a little more easy (all grunt)

**Comment by Lochemage** [comment]
@mc007 

Oh yes, and about wcPlay. It was the same for me, I was looking for that kind of tool but couldn't find one that matched what I was looking for, so I wrote it myself. Let me know how you like it, especially after you have tried to create your own node types, that I think is the most important.

**Comment by Lochemage** [comment]
@mc007 

Had some time to look at your latest updates, nice work on the documentation so far!

I had some trouble installing via bower, dependency "almost" doesn't seem to exist.

The most pressing issue right now is "index.html" and "demo.js", those have to work without making any changes to them. Currently running index outputs these script errors:
- Uncaught ReferenceError: define is not defined
- demo.js:11 Uncaught ReferenceError: wcDocker is not defined

**Comment by mc007** [comment]
hi,
sorry, my fault. I fixed that all: 
- bower_components now all in the repo
- almost: fixed = its 'almond'

index.html is untouched except some markup fixes
demo.js has as said still a delay: see open issues.

I am looing into the iframe issue now.

**Comment by Lochemage** [comment]
@mc007 

Okay, I see that index runs fine now, however, it no longer works after I run the automated build process that generates all the files in the build folder. Is there some other way you are building the files now?

**Comment by mc007** [comment]
iframe issue: fixed calls docker(this._panel) in Base. 

testing build, one moment

**Comment by mc007** [comment]
ok, forgot a file to add, works now. sorry,

builds however:

sudo npm install
grunt build 

that creates api docs and builds all code and themes (not using anymore your scripts). I will complete this in the change log a little better.

**Comment by mc007** [comment]
just one issue left (apart from some incomplete jsdoc changes): 

demo.js (used by index.html) has a delay, since wcDocker doesn't seems to be not yet in global, regardless wcDocker.min.js is pulled before demo.js = no idea.

be aware, indexDev.html is using Code/samples/demoAMD.js (doesn't need a delay) 

not sure how to fix it, but I'd say it can wait a little.

**Comment by Lochemage** [comment]
@mc007 

Ah okay, it builds the source files, but I am now getting an error when it tries to build the docs:
- Running "jsdoc:dist" (jsdoc) task
- Warning: jsdoc terminated with a non-zero exit code Use --force to continue.
and a popup message that shows:
node_modules\jsdoc\jsdoc.js Line 1, Char 1, Invalid Character, Code 880A03F6, Source Microsoft JScript compilation error

Also, with this change I can remove the Compiler folder, but how about builder.js? is that now replaced with code/ThemeBuilder? Are you using createDocs.sh/2?

**Comment by mc007** [comment]
jsdoc: seems an issue with windows i guess, let me check. (being on linux here with node 10.x)

compiler folder: needs still your jsdoc config; also search the Gruntfile.js with any entries for 'Compiler/'

g

**Comment by Lochemage** [comment]
@mc007 

Not sure why, though. I use the exact same library with my old doc build process and it worked fine. Perhaps something to do with grunt?  It does say that you can use the --force option to ignore that error, maybe that will work?

**Comment by mc007** [comment]
themebuilder:
index.html : is pulling it as normal js file, 
indexDev.html : yes, its using an AMD module as well. 

I did that because i didn't want to increase the size if wcDocker.min.js

createDocs.sh | createDocs2.sh : can be deleted now too.

jsdoc: nope, --force does't help here. there is indeed an issue parsing some stuff. let me check

**Comment by mc007** [comment]
jsdoc crash: confirmed, its an issue on windows. I will investigate this this evening.

**Comment by Lochemage** [comment]
@mc007 

Weird, wonder why it works fine on windows when I run it using Node...

**Comment by mc007** [comment]
yeah, i just did run Compiler/compile_documents and all is fine. weird, looks indeed like the grunt plugin is doing the crash.

**Comment by mc007** [comment]
ok, i will remove the grunt-jsdoc task for now til this is fixed, til then please proceed with your batch file.

**Comment by Lochemage** [comment]
@mc007 

Yeah, as much as it would have been awesome to have it all together, I might just need to keep the old stuff around. Oh well. Perhaps I can clean it up a bit by putting the grunt build stuff into the Compile folder, so they would both be together.

Anyway, I'm up way too late and I have work tomorrow, night!

**Comment by Lochemage** [comment]
@mc007 

Oh, and great job as well! Once everything is ready, I'll put it through some real testing and then pull it all into my repo!

**Comment by mc007** [comment]
ok, found the problem with the grunt-plug, its calling jsdoc wrongly. i could fix that my self in a fork but I'd prefer to report it just and then wait.

**Comment by mc007** [comment]
hi, and before you pull it into your repo, may be you create branch of master before, same for the changes as well, so your users can test it before :-)

ok, off for today.
g

**Comment by mc007** [comment]
hi,
what's missing for being 'ready'  ? There is another PR with API improvements and extensions (optional).
thanks,
g

**Comment by Lochemage** [comment]
@mc007 

I haven't had a chance to dig into it this weekend, I found myself swamped fixing issues with another project.

As far as my previous tests go, things seem to be working for the most part. My next course of action is really just nit-picking, code formatting, cleaning up code and files, etc.

I'll try to get to it soon, though.

**Comment by mc007** [comment]
ok, I will go on adding that API directly, should work all backward compatible anyways(fer is comming).

you have any link to your main project(s)?

thanks, don't bother to much ;-) 
g

**Comment by Lochemage** [comment]
@mc007 

Unfortunately no, I'm talking about a project I'm working on for the company I work for. We have all our repos hosted privately in-house. It's all very hush hush with lots of NDA's and such...

**Comment by Lochemage** [comment]
@gbaumgart 

I've made a number of improvements to your fork and have pulled it into the master branch!

**Comment by gbaumgart** [comment]
great! lets go for PR-2: 
- some API improvements (more making use of DCL)
- a theme based on https://wrapbootstrap.com/theme/light-blue-responsive-admin-web-app-WB0T41TX4 
- some additions dealing with tabs and more tab-actions (minimize,...)

should be ready in 2-3 weeks or so; my goal is to fix this mess in our app: https://cloud.githubusercontent.com/assets/1205100/9825503/fc8c7ebc-58d3-11e5-8492-b5e026552003.png

thanks, we love wcDocker :-)

**Comment by Lochemage** [comment]
@gbaumgart 

Nice, looking forward to it.  I know a few people have suggested minimizing and maximizing before.

For now, I will close this issue. We can start another one when we are ready to implement those features.
